prov/util: Adding a hooks mechanism

nikitaxgusev opened the pull request.

Commit:
core/util: Adding MR cache hooking mechanism

1) Inserted the main part of hooks hadnling:
   Created util_mem_hooks.c and ofi_mr_hooks.h
2) Next steps are:
   2.1 Being ready with MMAP,MUNMAP and etc. calls
   2.2 Adding hooks part to the main side of ofi_mem_monitor

Signed-off-by: Nikita Gusev <nikita.gusev@intel.com>

Commit:
core/util: Working on Sean's comments

1) Added licenses to new files
2) Delete comments in ofi_mem_hooks.h
3) Renamed files
4) Dropped 'linux' prefix in names functions/vars
5) Took the ofi_get_page_size() func in libfaric src
6) Threw away SIZE_OF_VOID_P
7) Threw away while() and make a label for unlock in patch_symbol()
8) Removed _t prefix
9) Removed typedefs: But I got some warnings, while it is compiling.

Signed-off-by: Nikita Gusev <nikita.gusev@intel.com>

Commit:
prov/util: Continue to integrate mr cache

1. Added draft for configure.ac in order to define if we can use hooks
   mechanism. (WIP)
2. Move struct ofi_patcher_dl_iter_contex to ofi_mem_hooks.h. Just
   declare it util_mem_hooks.c
3. Modify ofi_patcher_base_module:
   3.1. Threw away .patch_symbol = ofi_patcher_patch_symbol from struct
   ofi_patcher_base_module{..};
   3.2. Now we use ofi_patcher_patch_symbol() for all calls (MMAP,MUNMAP
   and etc)
4. Adding hooks mechanism to a monitor core:
   4.1. Added patcher init struct struct ofi_patcher{..};
   4.2. ofi_patcher_init() - implement, ofi_patcher_cleanup() - empty right now,
   ofi_patcher-init():
       4.2.1. Define subscribe/unsubscribe functions
       4.2.2  Calling ofi_patcher_handler, where we do dlopen, handle SYS_MMAP,
       SYS_MUNMAP and etc
5. Modify ofi_monitor_add_cache(): adding a new branch to handle `monitor == patcher_monitor`
   then we call ofi_patcher_init()
6. Added memory_patcher_syscall to _intercept_mmap, _intercept_munmap
   besides original_*. WIP.Have questions.
7. Code refactoring:
   add/change comments

Signed-off-by: Nikita Gusev <nikita.gusev@intel.com>

shefty commented on prov/util/src/util_mem_hooks.c:
```diff
@@ -0,0 +1,572 @@
+	/*not implemented yet*/
+	/*if ((flags & MAP_FIXED) && (start != NULL)) {
+		opal_mem_hooks_release_hook (start, length, true);
+	}*/
```
The OMPI calls that intercept the memory calls (mmap, munmap, mremap, etc.), should all have a call to opal_mem_hooks_release_hook().  We will want to replace those with something like this:
```
ofi_patcher_handler(const void *addr, size_t len);
```
and define ofi_patcher_handler as:
```
void ofi_patcher_handler(const void *addr, size_t len)
{
    fastlock_acquire(&patcher.lock);
    ofi_monitor_notify(&patcher.monitor, addr, len);
    fastlock_release(&patcher.lock);
}
```
That will result in flushing any cached registrations that cover 'addr' + 'len'.

shefty commented on prov/util/src/util_mem_monitor.c:
```diff
@@ -341,3 +362,74 @@ void ofi_uffd_cleanup(void)
+		if (!ofi_patcher_unregister(addr, len, page_sizes[i]))
+			break;
+	}
+}
```
Because patcher intercepts all calls that may modify the virtual address to physical address mappings, the subscribe/unsubscribe calls aren't needed.  They can be no-ops.  Subscribe can just return 0.

nikitaxgusev replied:
Done it for now.

Commit:
prov/rxd: the patcher is no-ops mechanism

1. Added core_monitor variable to test
2. Threw away defines in order for fast testing
3. Defined ofi_pathcer_handler using in SYS_MMAP, SYS_MUNMAP and etc.
4. Code refactoring

Signed-off-by: Nikita Gusev <nikita.gusev@intel.com>

shefty commented on include/ofi_mr.h:
```diff
@@ -196,6 +209,7 @@ struct ofi_mr_cache_params {
 	size_t				max_cnt;
 	size_t				max_size;
 	int				merge_regions;
+	int				core_monitor;
```
This isn't quite what I had in mind.  I was thinking of defining:
```
extern struct ofi_mem_monitor *default_monitor;
```
'default' is probably a better name here.  Then in util_mem_monitor.c:
```
#if HAVE_UFFD_UNMAP
struct ofi_mem_monitor *default_monitor = &uffd.monitor;
#else
struct ofi_mem_monitor *default_monitor = &patcher_monitor.monitor'
```
An environment variable would allow us to change the default.  Though I would change based on a string, such as "userfaultfd" or "memoryhooks".

shefty commented on include/ofi_mem_hooks.h:
```diff
@@ -0,0 +1,65 @@
+	struct ofi_patcher_patch *patch;
+	bool remove;
+	int status;
+};
```
You mentioned in an email trying to combing all of the structures into one.  I agree that's a good idea, but let's get this working and merged first before bothering with too much cleanup.  Cleanup patches can follow once we have something working upstream.  Personally, I want to rename everything too, because 'patcher' is kind of goofy.

shefty commented on prov/util/src/util_mem_hooks.c:
```diff
@@ -0,0 +1,391 @@
+	void *result = _intercept_mmap (start, length, prot, flags, fd, offset);
+
+	return result;
+}
```
nit: I have no idea why the original code defines calls like intercept_xxx and __intercept_xxx.  These should be combined into a single function.  This is cleanup that can be done post merge, however.

nikitaxgusev replied:
@shefty They say [this](https://github.com/open-mpi/ompi/blob/master/opal/mca/memory/patcher/memory_patcher_component.c#L104-#L108), but I think, it is about for IBM environment,  we should drop this.

shefty replied:
I agree, we should drop this.

shefty commented on prov/util/src/util_mem_hooks.c:
```diff
@@ -0,0 +1,391 @@
+	int result = _intercept_munmap (start, length);
+
+	return result;
+}
```
There is a fabtests, fi_unmap_mem, that can be used to verify that the intercepts and mr cache are working.  That program will call mmap/munmap in between transfers in order to change the virtual to physical address translations.

shefty commented on prov/util/src/util_mem_monitor.c:
```diff
@@ -33,19 +33,28 @@
+	if (cache_params.core_monitor == 1) {
+		fastlock_init(&patcher_monitor->lock);
+		dlist_init(&patcher_monitor->list);
+	}
```
We will want to initialize all monitors that we can here, regardless of which one is selected as the default.

shefty commented on prov/util/src/util_mem_monitor.c:
```diff
@@ -66,19 +75,30 @@ void ofi_monitor_init(void)
 
+	fi_param_define(NULL, "core_monitor", FI_PARAM_BOOL,
+			"TODO: Need the description here. Choose core monitor");
+
```
Drop this for now.  We can manually set the default for testing and add a parameter to set the default after merging.

shefty commented on prov/util/src/util_mem_monitor.c:
```diff
@@ -66,19 +75,30 @@ void ofi_monitor_init(void)
+		/* We need to define the logic the choice.*/
+		assert(dlist_empty(&patcher_monitor->list));
+		fastlock_destroy(&patcher_monitor->lock);
+	}
```
Likewise, we should cleanup all monitors, regardless of the default.

Commit:
prov/util: code refactoring & added new env. var

1. Set default monitor in order to choose which we can aplly for the
system we have.

2. Defined FI_MR_CORE_MONITOR without functionality
   Plan is allowing us to change the default monitor

3. Code cleanup

Signed-off-by: Nikita Gusev <nikita.gusev@intel.com>
Signed-off-by: Mikhail Khalilov <mikhail.khalilov@intel.com>
